### PR TITLE
[SYCL-MLIR][mlir-translate] Set call attributes following callee

### PR DESCRIPTION
**Pull request by victor-eds**
Make translated call instructions attributes the same as those in the callee function.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp [hunk 1]
```diff
@@ -196,6 +196,7 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
           moduleTranslation.lookupFunction(attr.getValue());
       call = builder.CreateCall(callee, operandsRef);
       call->setCallingConv(callee->getCallingConv());
+      call->setAttributes(callee->getAttributes());
```
Comment (sommerlukas) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:199]: Is it a requirement that call site and function attributes always match? Or is there any meaningful and legal use-case in which the two attribute sets could differ (beyond the call site only attributes)?

File: polygeist/tools/cgeist/Test/Verification/sycl/usm-wrapper.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// RUN: clang++ -fsycl -fsycl-device-only -S -emit-llvm -o - -fsycl-targets=spir64-unknown-unknown-syclmlir -O3 %s | FileCheck %s
+
```
Comment (sommerlukas) [inline polygeist/tools/cgeist/Test/Verification/sycl/usm-wrapper.cpp:2]: Can we document the purpose of this test in a short comment?

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:199]: AFAIK, there are some cases, e.g., calls with attributes like `always_inline`.

This is just a work-around the lack of proper `llvm.call` attributes support. When this is upstreamed, we may want to do the proper thing and make attributes first-class citizens in the dialect. For now, there is no standard way to do so.

We may want these to come from codegen, but doing so would be too much work:

- Generating in codegen (possibly cheap if reusing what we have for functions)
- `func.call -> llvm.call` conversion
- `llvm.call -> LLVM call instruction`

Two last steps are what we may wanna upstream.

Only cases in which we are being incorrect now are for the cases in which func-site and call-site may differ, but I think this is at least forward progress.

Comment (sommerlukas) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:199]: Okay, makes sense. Maybe we should add a comment here explaining that this is not the final solution and may cause issues.

**Review by sommerlukas — APPROVED** [review]

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:199]: Good point :+1:

**Pull request merged**